### PR TITLE
Enable page list to expand in list view

### LINKED ISSUE #45482: Page List: Display items in the List View
As outlined in https://github.com/WordPress/gutenberg/issues/42257, the items inside Page List should be displayed in the List View:
Screenshot 2022-11-02 at 09 28 06

Previous art: https://github.com/WordPress/gutenberg/pull/39087

### PR DESCRIPTION
**Pull request by draganescu**
## What?

Closes https://github.com/WordPress/gutenberg/issues/45482

Adds the ability to the page list block to expand in the list view.

## Why?

To reflect on the hierarchical structure that we're representing, particularly when the page list is used inside a navigation block.

## How?

Add a new Page List Item block that can only appear as a chidl of Page List.

## Testing Instructions
- check out this PR
- be sure to have a few pages and some of them to have a parent
- in a post add a Page List block
- open List View 
- notice the page list block can be expanded and all its children appear

## Screenshots or screencast 

https://user-images.githubusercontent.com/107534/202382441-f0750e87-39bd-4fb6-aa67-f583021150f3.mp4

## Known issues:

- [x] While the inner blocks are locked they can be unlocked by the user. The problem is that any edits will not be saved, as the Page List block is always using the hierarchy of pages available, including titles for labels and page URL for links. We need to make an option for locks to be unlock-able or come up with something else.
- [x] The new inner blocks cause some layout issues when the page list block is used in a navigation block
- [x] The post is dirty after the page list renders because the innerblocks are added too late? (fixed in https://github.com/WordPress/gutenberg/pull/45776/commits/f69f59d5746163d9e89d7612c4caf5eac94f817d )

### CODE REVIEW

**Comment by jasmussen** [comment]
Looks good. 

<details>
<summary>
Here's normal items:
</summary>

Screenshot 2022-11-16 at 11 43 12

</details>
<details>
<summary>
Here's a page list:
</summary>

Screenshot 2022-11-16 at 11 44 02

</details>
<details>
<summary>
However I'm seeing the page list becoming a vertical list in the canvas in this PR:
</summary>

Screenshot 2022-11-16 at 11 44 21
</details>
<details>
<summary>
Here's the same menu in trunk:
</summary>

Screenshot 2022-11-16 at 11 45 39
</details>

File: packages/block-library/src/page-list/edit.js [hunk 1]
```diff
@@ -52,6 +51,34 @@ export default function PageListEdit( { context, clientId } ) {
...
+			const children = makeBlockTemplate( page.id );
+			item = [ 'core/page-list-item', pageProps, children ];
+
+			template.push( item );
+
+			return template;
+		}, [] );
+	};
+
+	const pagesTemplate = makeBlockTemplate();
```
Comment (Mamaduka) [inline packages/block-library/src/page-list/edit.js]: It would be a good idea to memoize the inner block template; otherwise, `useInnerBlockTemplateSync` will try to synchronize it on every rerender.

Comment (draganescu) [reply packages/block-library/src/page-list/edit.js]: Yes! 💯 👍🏻

**Comment by draganescu** [comment]
Discussing the bug making the post dirty with @talldan he pointed to how templated blocks work where the template is placed using replace blocks which is a persistent change:

https://github.com/WordPress/gutenberg/blob/fecc1d2aa66c2fb2c6379fb47de5d7cd71f01f8c/packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js#L84-L96

This implementation sets the template asynchronously which is after the initial render so the post is marked as dirty, as it should, but not in this case.

Dan suggested that to try and manually replace blocks instead of using a template. It remains to be seen which way is best.

File: packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js [hunk 2]
```diff
@@ -81,6 +82,7 @@ export default function useInnerBlockTemplateSync(
 			);
 
 			if ( ! isEqual( nextBlocks, currentInnerBlocks ) ) {
+				__unstableMarkNextChangeAsNotPersistent();
```
Comment (Mamaduka) [inline packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js]: I'm not sure if patching the `useInnerBlockTemplateSync` hook for a single use case is a good idea 😅

Comment (draganescu) [reply packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js]: It's not only this, we've hit this before: see https://github.com/WordPress/gutenberg/pull/45843

File: packages/block-library/src/page-list-item/block.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "core/page-list-item",
+	"title": "Page List Item",
+	"category": "design",
+	"parent": [ "core/page-list" ],
+	"description": "Display a list of all pages.",
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/block.json]: I think this is inaccurate. I think it should say something like:
```suggestion
	"description": "Displays page, inside a list of all pages .",
```

File: packages/block-library/src/page-list-item/block.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	"keywords": [ "menu", "navigation" ],
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/block.json]:
```suggestion
	"keywords": [ "page", "menu", "navigation" ],
```

File: docs/reference-guides/core-blocks.md [hunk 5]
```diff
@@ -431,6 +431,15 @@ Display a list of all pages. ([Source](https://github.com/WordPress/gutenberg/tr
 -	**Supports:** ~~html~~, ~~reusable~~
 -	**Attributes:** 
 
+## Page List Item
+
+Display a list of all pages. ([Source](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/page-list-item))
```
Comment (scruffian) [inline docs/reference-guides/core-blocks.md]:
```suggestion
Displays page, inside a list of all pages. ([Source](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/page-list-item))
```

File: packages/block-library/src/page-list-item/block.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "core/page-list-item",
+	"title": "Page List Item",
+	"category": "design",
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/block.json]: Page List is under widgets.

File: packages/block-library/src/page-list-item/block.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		"fontSize",
+		"customFontSize",
+		"showSubmenuIcon",
+		"style",
+		"openSubmenusOnClick"
+	],
+	"supports": {
+		"reusable": false,
+		"html": false,
+		"lock": false
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/block.json]: Should we add ` inserter: false`?

File: packages/block-library/src/page-list-item/editor.scss [hunk 8]
```diff
@@ -0,0 +1,59 @@
+.wp-block-navigation {
+	// Block wrapper gets the classes in the editor, and there's an extra div wrapper for now, so background styles need to be inherited.
+	.wp-block-page-list > div,
+	.wp-block-page-list {
+		background-color: inherit;
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/editor.scss]: From what I can see this is unnecessary.

**Comment by scruffian** [comment]
I assume you tried using the Page Link block?

File: packages/block-library/src/page-list-item/editor.scss [hunk 9]
```diff
@@ -0,0 +1,59 @@
+.wp-block-navigation {
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/editor.scss]: This file is already used in `page-list`, do we need another copy of it?

File: packages/block-library/src/page-list-item/index.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+export { metadata, name };
+
+export const settings = {
+	__experimentalLabel: ( { label } ) => label,
+	icon,
+	example: {},
+	edit,
+};
+
+export const init = () => initBlock( { name, metadata, settings } );
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/index.js:24]: Why have a separate file? Why not just call `init` here?

File: packages/block-library/src/page-list-item/style.scss [hunk 11]
```diff
@@ -0,0 +1,17 @@
+// The Page List block should inherit navigation styles when nested within it
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/style.scss]: Same question here about doubling up the styles.

**Comment by scruffian** [comment]
Thanks for working on this, it's looking great. I just left a few questions about things I didn't understand.

File: packages/block-library/src/page-list-item/editor.scss (hunk 9 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/editor.scss]: I think I need to remove some stuff from page list for these ones.

File: packages/block-library/src/page-list-item/block.json (hunk 3 above)
Comment (getdave) [reply packages/block-library/src/page-list-item/block.json]: Should it be

> Displays _a_ page...

File: docs/reference-guides/core-blocks.md (hunk 5 above)
Comment (getdave) [reply docs/reference-guides/core-blocks.md]: > Displays _a_ page...

Missing `a`

File: packages/block-library/src/page-list-item/block.json [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		"overlayBackgroundColor",
+		"customOverlayBackgroundColor",
+		"fontSize",
+		"customFontSize",
+		"showSubmenuIcon",
+		"style",
+		"openSubmenusOnClick"
+	],
+	"supports": {
+		"reusable": false,
```
Comment (getdave) [inline packages/block-library/src/page-list-item/block.json:44]:
```suggestion
	"supports": {
		"parent": [ "core/page-list" ],
		"reusable": false,
```

Also probably define the `parent` property so it can only be inserted on the Page List block.

**Review by getdave — COMMENTED** [review]
Some things I noticed.

File: packages/block-library/src/page-list-item/block.json (hunk 12 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/block.json:44]: This is set but I see that "parent" is not a child of supports, but a child of root which I did set.

File: packages/block-library/src/page-list-item/block.json (hunk 7 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/block.json]: I think this is a good idea.

File: packages/block-library/src/page-list-item/index.js (hunk 10 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/index.js:24]: This is a pattern across the `block-library` package. It comes from https://github.com/WordPress/gutenberg/pull/42258 which enabled easier importing of only one block.

File: packages/block-library/src/page-list-item/style.scss (hunk 11 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/style.scss]: I think it's safe to delete the ones in the new block.

File: packages/block-library/src/page-list-item/editor.scss (hunk 8 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/editor.scss]: It is.

File: packages/block-library/src/page-list-item/block.json [hunk 13]
```diff
@@ -3,10 +3,10 @@
 	"apiVersion": 2,
```
Comment (getdave) [inline packages/block-library/src/page-list-item/block.json:3]: Does this need the `parent` to be set to `core/page-list`?
```suggestion
	"apiVersion": 2,
	"parent": [ "core/page-list" ],
```

**Review by getdave — COMMENTED** [review]
As things stand I don't think this is working 😞 

<details>
https://user-images.githubusercontent.com/444434/203301672-d0950170-a1b5-48cb-9d3e-f19384309548.mp4
</details>

I cleared all my menus then added a new Nav block.

- a new Nav was automatically created
- I saw a Page List in the offcanvas
- I tried clicking on Page List in the offcanvas but I couldn't seem to select it
- I went to the main List View and clicked on Page List. I _was_ able to selected it.

Basically everything seems out of sync as per the video.

File: packages/block-library/src/page-list-item/block.json (hunk 13 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/block.json:3]: Isn't this set a few lines down?

**Comment by draganescu** [comment]
Thanks for the testing @getdave 🙇🏻 

- the off canvas editor component specifically disables block selection on click so that we dont immediately switch to that block's inspector, hence the introduction of the pencil icon. 
- this PR I believe shows a bug in the Navigation block's assumptions on when to automagically create a navigation: we fall back to a page list when no menus exist, but here selecting the page list alone (not editing) triggers the autosave, which I think now it's because it has innerBlocks, and we go the hasUncontroledInnerBlocks route.

None seem to be problems of this PR.

File: packages/block-library/src/page-list-item/block.json (hunk 13 above)
Comment (getdave) [reply packages/block-library/src/page-list-item/block.json:3]: Ok. Maybe my diff didn't show that 👍

File: packages/block-library/src/page-list-item/edit.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+export default function PageListItemEdit( { context, attributes } ) {
+	const { id, label, link, hasChildren } = attributes;
+	const isNavigationChild = 'showSubmenuIcon' in context;
+	const frontPageId = useFrontPageId();
+	return (
+		<li
+			key={ id }
+			className={ classnames( 'wp-block-pages-list__item', {
+				'has-child': hasChildren,
+				'wp-block-navigation-item': isNavigationChild,
```
Comment (scruffian) [inline packages/block-library/src/page-list-item/edit.js:42]: It seems like there is quite a lot of overlap between this file and the navigation link - they need to share the same layout. Is there anything we can do to ensure they stay in sync?

**Review by scruffian — APPROVED** [review]
This is testing well for me. I have a few small concerns about sharing the layout with the navigation-link block, but nothing that should stop you merging.

File: packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js (hunk 2 above)
Comment (Mamaduka) [reply packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js]: Can we remove this from this PR and wait on the decision in https://github.com/WordPress/gutenberg/pull/45843?

File: packages/block-library/src/page-list-item/edit.js (hunk 14 above)
Comment (getdave) [reply packages/block-library/src/page-list-item/edit.js:42]: I _think_ this has been an issue in `core/page-list` for a while. The problem has now just moved to the `item`. Am I right @draganescu ?

**Comment by getdave** [comment]
> * the introduction of the pencil icon.

As long as that works for the Page List (parent) block itself.

> here selecting the page list alone (not editing) triggers the autosave

The code for this autosave was written specifically to check for a dirty state on the blocks that are pased into the component but it is a little bit "hacky". 

Did you create an Issue to track this bug? I think that's important 🙏

File: packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js (hunk 2 above)
Comment (draganescu) [reply packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js]: I'll try to rebase on that or something. b/c otherwise this PR does not work.

**Comment by getdave** [comment]
What's the status on this one? Blocked due to the [sync problems with inner blocks](https://github.com/WordPress/gutenberg/pull/45776#discussion_r1024982643)?

File: packages/block-library/src/page-list-item/edit.js (hunk 14 above)
Comment (draganescu) [reply packages/block-library/src/page-list-item/edit.js:42]: Yes. Page list item is mostly a copy of the old PageItems component used inside the page list block.

**Comment by draganescu** [comment]
Yes, @getdave @scruffian should we merge with the known bug that adding a page list makes the post dirty? Should we wait for  https://github.com/WordPress/gutenberg/pull/45843 ?

File: packages/block-library/src/page-list-item/edit.js (hunk 14 above)
Comment (scruffian) [reply packages/block-library/src/page-list-item/edit.js:42]: I created a follow up issue for this here: https://github.com/WordPress/gutenberg/issues/46050

**Review by scruffian — APPROVED** [review]
This is looking good. I think we should deal with the edit experience in a followup: https://github.com/WordPress/gutenberg/issues/46051
